### PR TITLE
fix: respect `fd_prestat_dir_name` `path_max_len` argument

### PR DESCRIPTION
**Pull request by rvolosatovs**
Refs https://github.com/bytecodealliance/wasmtime/pull/6440#discussion_r1202998079

### CODE REVIEW

**Review by jameysharp — APPROVED** [review]
I think this is right but I'd like @pchickey to take a look.

File: crates/test-programs/wasi-tests/src/bin/dir_fd_op_failures.rs [hunk 1]
```diff
@@ -5,6 +5,17 @@ unsafe fn test_fd_dir_ops(dir_fd: wasi::Fd) {
...
+    let fd = (3..)
+        .map_while(|fd| {
+            wasi::fd_prestat_get(fd)
+                .ok()
+                .map(|wasi::Prestat { tag, .. }| (fd, tag))
+        })
+        .find_map(|(fd, tag)| (tag == wasi::PREOPENTYPE_DIR.raw()).then_some(fd))
+        .expect("failed to find preopen directory");
+    let r = wasi::fd_prestat_dir_name(fd, [].as_mut_ptr(), 0);
+    assert_eq!(r, Err(wasi::ERRNO_NAMETOOLONG));
```
Comment (pchickey) [inline crates/test-programs/wasi-tests/src/bin/dir_fd_op_failures.rs:17]: Can we also add a case where we pass a buffer which is 1 byte longer than required

File: crates/test-programs/wasi-tests/src/bin/dir_fd_op_failures.rs [hunk 2]
```diff
@@ -5,6 +5,18 @@ unsafe fn test_fd_dir_ops(dir_fd: wasi::Fd) {
...
+    let (pr_fd, pr_name_len) = (3..)
+        .map_while(|fd| wasi::fd_prestat_get(fd).ok().map(|stat| (fd, stat)))
+        .find_map(|(fd, wasi::Prestat { tag, u })| {
+            (tag == wasi::PREOPENTYPE_DIR.raw()).then_some((fd, u.dir.pr_name_len))
+        })
+        .expect("failed to find preopen directory");
+    let r = wasi::fd_prestat_dir_name(pr_fd, [].as_mut_ptr(), 0);
+    assert_eq!(r, Err(wasi::ERRNO_NAMETOOLONG));
+
+    let r = wasi::fd_prestat_dir_name(pr_fd, vec![0; pr_name_len].as_mut_ptr(), pr_name_len + 1);
```
Comment (pchickey) [inline crates/test-programs/wasi-tests/src/bin/dir_fd_op_failures.rs:17]:
```suggestion
    // Test that passing a larger than necessary buffer works correctly
    let r = wasi::fd_prestat_dir_name(pr_fd, vec![0; pr_name_len + 1].as_mut_ptr(), pr_name_len + 1);
```

Comment (pchickey) [reply crates/test-programs/wasi-tests/src/bin/dir_fd_op_failures.rs:17]: I'd rather not risk this being UB in the case of an incorrect implementation

**Review by pchickey — APPROVED** [review]

File: crates/test-programs/wasi-tests/src/bin/dir_fd_op_failures.rs [hunk 3]
```diff
@@ -5,6 +5,24 @@ unsafe fn test_fd_dir_ops(dir_fd: wasi::Fd) {
...
+        })
+        .expect("failed to find preopen directory");
+
+    let r = wasi::fd_prestat_dir_name(pr_fd, [].as_mut_ptr(), 0);
+    assert_eq!(r, Err(wasi::ERRNO_NAMETOOLONG));
+
+    // Test that passing a larger than necessary buffer works correctly
+    let r = wasi::fd_prestat_dir_name(
+        pr_fd,
+        vec![0; pr_name_len + 1].as_mut_ptr(),
```
Comment (alexcrichton) [inline crates/test-programs/wasi-tests/src/bin/dir_fd_op_failures.rs:21]: I believe that this as-written is ok, but I'd recommend avoiding this pattern in the future as it's so easy to get wrong. This only works if the `vec![...]` allocate here lives longer than the function call which I think rustc's temporary rules satisfy here, but I at least personally think it'd be more clear to move the vec above this statement to ensure it lives across the statement.

Comment (rvolosatovs) [reply crates/test-programs/wasi-tests/src/bin/dir_fd_op_failures.rs]: Thanks! Yes, good point, I would guess that `miri` should complain about this or at least I remember seeing it fail on a similar pattern in a different project before.

**Comment by rvolosatovs** [comment]
@alexcrichton @pchickey please take a look. I've addressed the last bit of feedback from @alexcrichton.
Previous merge attempt has failed due to `poll_oneof_files` failure on macos, which appears to be a flaky test

**Review by alexcrichton — APPROVED** [review]

**Comment by alexcrichton** [comment]
Yeah that sounds familiar that that test is flaky, it's pretty rare but one we've never managed to track down.

**Pull request merged**